### ISSUE TITLE
focus proxy function

### ISSUE THREAD

**Issue by Tixierae** [juba/scatterD3#76]
Would it be possible to create a function like `visFocus` of `visNetwork`? (see `Modify your network with visNetworkProxy` here: https://datastorm-open.github.io/visNetwork/shiny.html) 

after a `scatterD3` object has been created, the function would allow to zoom on a specific point, something like:
```r
my_scatterD3 %>% Focus(label_or_coordinates, scale=4)
```
I think this would be a very nice feature.

**Comment by juba** [comment]
It's a good idea, even if I'm not sure I could find time to implement it. However I see two ways of doing this : 

- only once, at startup, when your plot is loaded a zoom is automatically triggered on a given point at a given level
- dynamically, within shiny. An input could list points and when one is selected a zoom is triggered on it at a given level

Which one were you thinking of ?

**Comment by Tixierae** [comment]
I was thinking of the second one. The user would select a point from a dropdown (such as a selectizeInput) and the view would automatically focus on the point (either directly given the label of the point, or from its coordinates)

**Comment by juba** [comment]
Ok, this one took longer than expected, but I think it's working.

You can take a look at the sample shiny app for an example of the feature, and at its source code to see how it works : https://github.com/juba/scatterD3_shiny_app

**Comment by Tixierae** [comment]
Amazing, thank you very much! I checked your example app online and it works very well. However, when I install the latest version locally with `devtools::install_github("juba/scatterD3")` and try to use `scatterD3()` with this new feature, I get an error `unused arguments (zoom_on = zoom_on, zoom_on_level = 8)`. However, the `zoom_on` and `zoom_on_level` parameters do appear when I query the documentation with `?scatterD3`

Note: when I install from GitHub, I receive a warning that I don't think I was receiving before (I last installed from GitHub a few weeks ago). I haven't changed my configuration, so maybe something has changed on your side?

```
WARNING: Rtools is required to build R packages, but no version of Rtools compatible with R 3.4.4 
was found. (Only the following incompatible version(s) of Rtools were found:3.3)
```

**Comment by juba** [comment]
Hmm that's strange. I just tried to reinstall the github version and it seems to work.

I don't think you should need Rtools to install `scatterD3`. Maybe you can try to upgrade it to make the warning go, but I'm not sure it will fix the problem.

**Comment by Tixierae** [comment]
I removed and re-installed the package, and I'm not receiving the error anymore.

**Comment by Tixierae** [comment]
However, I'm still experiencing issues. In the app below, the plot disappears when everything other than `None` is selected from the dropdown. Maybe I'm missing something?

```r
library(shiny)
library(scatterD3)

mtcars$names = rownames(mtcars)

ui = shinyUI(
  fluidPage(
    fluidRow(
      selectInput("scatterD3_zoomon", "Zoom on :", choices = c("None", rownames(mtcars))),
      br()
    ),
    fluidRow(
      scatterD3Output('scatterd3')
    )
  )
)
server = shinyServer(function(input, output, session) {

  output$scatterd3 = renderScatterD3({
    
    zoom_on <- if (input$scatterD3_zoomon == "None") {
      NULL 
    } else {
      as.numeric(mtcars[input$scatterD3_zoomon, c('wt','mpg')])
    }
    
    scatterD3(data = mtcars, x = wt, y = mpg, lab = names,
              col_var = cyl, symbol_var = am,
              xlab = "Weight", ylab = "Mpg", col_lab = "Cylinders",
              symbol_lab = "Manual transmission",zoom_on=zoom_on,zoom_on_level=8)
  })
})
shinyApp(ui=ui, server=server)

```

**Comment by juba** [comment]
Yes, you're right, I get an error too. Thanks for reporting, I'll look into it.

**Comment by Tixierae** [comment]
No prob. Are you getting an error? In my case, it's just that the plot silently disappears, but the app keeps running and the plot re-appears if I re-select `None` in the dropdown.

**Comment by juba** [comment]
Yes, same thing here. It's because the new code is causing a JavaScript error which stops the plot from being created.

**Comment by juba** [comment]
Ok, small update, it does work if you add `transitions = TRUE` to your `scatterD3` call.

**Comment by Tixierae** [comment]
Indeed, it does the trick, thank you! I noticed though that the first time you zoom (initially, or after resetting the zoom), it zooms into an empty area of the plot, and then it goes to the selected element. After this, all subsequent selections are handled perfectly.

**Comment by juba** [comment]
Yes, you're right. I know this is suboptimal, but unfortunately I didn't find a way to make it work otherwise. Even translating before zooming would be better, but I can't manage to make it right in this order...

**Comment by Tixierae** [comment]
No prob, it's still a very useful feature. Thanks!

**Comment by Tixierae** [comment]
I noticed another issue, hopefully this one will be easier to fix. When custom colors (e.g., `colors=my_cols` below) are passed, the zoom feature stops working. See the reproducible example below:

```r
library(shiny)
library(scatterD3)

mtcars$names = rownames(mtcars)
my_cols = c("#FF0000","#FF4000","#FF8000","#FFBF00","#FFFF00","#FFFF80")

ui = shinyUI(
  fluidPage(
    fluidRow(
      selectInput("scatterD3_zoomon", "Zoom on :", choices = c("None", rownames(mtcars))),
      br()
    ),
    fluidRow(
      scatterD3Output('scatterd3')
    )
  )
)
server = shinyServer(function(input, output, session) {
  
  output$scatterd3 = renderScatterD3({
    
    zoom_on <- if (input$scatterD3_zoomon == "None") {
      NULL 
    } else {
      as.numeric(mtcars[input$scatterD3_zoomon, c('wt','mpg')])
    }
    
    scatterD3(data = mtcars, x = wt, y = mpg, lab = names,
              col_var = cyl, symbol_var = am,colors=my_cols,
              xlab = "Weight", ylab = "Mpg", col_lab = "Cylinders",
              symbol_lab = "Manual transmission",zoom_on=zoom_on,zoom_on_level=8,
              transitions=TRUE)
  })
})
shinyApp(ui=ui, server=server)
```

**Comment by juba** [comment]
So, this last bug should now be fixed, and the zoom movement when zooming on a point should be better (translation before scaling). There's still a problem with scaling when the plot has been resized, though.

Thanks again for the report and the reproducible example.

**Comment by Tixierae** [comment]
Apologies for the late reply - I have been too busy lately to check out your fix. I just did, and everything works perfectly now. The transition is much smoother, and the colors show just fine. Thank you very much, this is really great!
